clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsLoadBalancerProvider.java

```diff
@@ -92,27 +106,116 @@ public String getCloudProvider() {
-    // which is not even implemented yet
+    // unused method
+    throw new UnsupportedOperationException("Not implemented.");
   }
 
   @Override
-  public Set<AmazonLoadBalancer> getApplicationLoadBalancers(String application) {
-    return null; // TODO - Implement this.  This is used to show load balancers and reveals other
-    // buttons
-  }
-
-  private String getEcsAccountName(String awsAccountName) {
-    for (ECSCredentialsConfig.Account ecsAccount : ecsCredentialsConfig.getAccounts()) {
-      if (ecsAccount.getAwsAccount().equals(awsAccountName)) {
-        return ecsAccount.getName();
+  public Set<EcsLoadBalancer> getApplicationLoadBalancers(String application) {
+    // Find the load balancers currently in use by ECS services in this application
+    Set<Service> services =
+        ecsServiceCacheClient.getAll().stream()
+            .filter(service -> service.getApplicationName().equals(application))
+            .collect(Collectors.toSet());
+    log.debug("Retrieved {} services for application '{}'", services.size(), application);
+
+    Collection<String> allTargetGroupKeys = ecsTargetGroupCacheClient.getAllKeys();
+    log.debug(
+        "Retrieved {} target group keys for application '{}'",
+        allTargetGroupKeys.size(),
+        application);
+
+    Map<String, Set<String>> targetGroupToServicesMap = new HashMap<>();
+    Set<String> targetGroupKeys = new HashSet<>();
+
+    // find all the target group cache keys
+    for (Service service : services) {
+      String awsAccountName =
+          ecsAccountMapper.fromEcsAccountNameToAwsAccountName(service.getAccount());
+      for (LoadBalancer loadBalancer : service.getLoadBalancers()) {
+        if (loadBalancer.getTargetGroupArn() != null) {
+          String tgArn = loadBalancer.getTargetGroupArn();
+          String keyPrefix =
+              String.format(
+                  "%s:%s:%s:%s:%s:",
+                  AmazonCloudProvider.ID,
+                  TARGET_GROUPS.getNs(),
+                  awsAccountName,
+                  service.getRegion(),
+                  ArnUtils.extractTargetGroupName(tgArn).get());
+          Set<String> matchingKeys =
+              allTargetGroupKeys.stream()
+                  .filter(key -> key.startsWith(keyPrefix))
+                  .collect(Collectors.toSet());
+          targetGroupKeys.addAll(matchingKeys);
+          // associate target groups with services it contains targets for
+          if (targetGroupToServicesMap.containsKey(tgArn)) {
+            Set<String> serviceList = targetGroupToServicesMap.get(tgArn);
+            serviceList.add(service.getServiceName());
+            targetGroupToServicesMap.put(tgArn, serviceList);
+          } else {
+            Set<String> srcServices = new HashSet<>();
+            srcServices.add(service.getServiceName());
+            targetGroupToServicesMap.put(tgArn, srcServices);
```

Comment:
Since this logic is very similar, it might make more sense to move the common logic out of the if-else?

Reply:
actually I can make the `else` block more concise, that'll differentiate them